Add converters to DatePicker and TimePicker for DateOnly and TimeOnly

symbiogenesis opened the pull request:
I realize that full DateOnly and TimeOnly support is not implemented yet. This is a start in that direction.

Fixes https://github.com/dotnet/maui/issues/20438

As mentioned in the comments there, fixing that bug is merely a subset of https://github.com/dotnet/maui/issues/1100 which can be fixed without the need for a full implementation.

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
There are some compilation errors:

`D:\a\_work\1\s\src\Controls\src\Core\TimePicker\TimePicker.cs(71,47): error CS0246: The type or namespace name 'TimeOnlyConverter' could not be found (are you missing a using directive or an assembly reference?) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.19041.0]`

symbiogenesis committed:
add converters for DateOnly and TimeOnly

symbiogenesis commented:
> There are some compilation errors

Should be fixed now

symbiogenesis commented:
@Eilon

symbiogenesis commented on src/Controls/src/Core/TimePicker/TimePicker.cs:
```diff
@@ -67,6 +67,9 @@ public double CharacterSpacing
 		}
 
 		/// <include file="../../docs/Microsoft.Maui.Controls/TimePicker.xml" path="//Member[@MemberName='Time']/Docs/*" />
+#if NET6_0_OR_GREATER
```
It was explained to me in another PR that netstandard2.0 support will be maintained for the foreseeable future

https://github.com/dotnet/maui/pull/19965#event-13464335155

symbiogenesis commented:
This one would be nice to have. @Foda

Foda commented:
> This one would be nice to have. @Foda

A simple unit test would be great!

symbiogenesis commented:
> @symbiogenesis are you still interested in completing this one? Else I will take over!

Looks like I had some changes locally that I didn't push, yet. So I just pushed those.

Needs the conflicts resolved. If you could resolve those, I would be appreciative.

jfversluis commented:
I don't think adding the properties was part of the plan? The whole point of this was that it would convert TimeOnly and DateOnly to TimeSpan & DateTime right? So that we could use this with the existing properties.

As mentioned in #27930 and #27921 unfortunately we can't use DateOnly and TimeOnly since they are not supported in netstandard. And as you pointed out yourself, we're going to have to support that for a while to come. So this seems like a nice in-between solution :)

symbiogenesis committed:
add tests

symbiogenesis commented:
Ok, that makes sense. Must be why I didn't include those earlier.

Added some tests.

jfversluis committed:
Retarget to net10

jfversluis commented:
/azp run MAUI-public

jfversluis committed:
Fix tests

rmarinho commented:
/azp run

rmarinho reviewed (changes requested):
Seems some unit tests are failing 
<img width="1073" alt="Screenshot 2025-04-22 at 10 26 54" src="https://github.com/user-attachments/assets/e1096e98-e02e-485a-b703-aeb63a5aff1f" />

rmarinho commented:
Needs rebase now

PureWeen reviewed (changes requested):
@StephaneDelcroix review and approve before we merge please

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace Microsoft.Maui.Controls;
+    
+public class DateOnlyToDateTimeConverter : TypeConverter, IExtendedTypeConverter
```
rename to DateTimeTypeConverter

StephaneDelcroix commented on src/Controls/src/Core/TimeOnlyToTimeSpanConverter.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace Microsoft.Maui.Controls;
+
+public class TimeOnlyToTimeSpanConverter : TypeConverter, IExtendedTypeConverter
```
rename to TimeSpanTypeConverter

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+        if (DateTime.TryParse(value, CultureInfo.InvariantCulture, DateTimeStyles.None, out var result))
+        {
+            return DateOnly.FromDateTime(result);
```
return value should ba a DateTime

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+        if (value is DateTime dateTime)
+        {
+            return DateOnly.FromDateTime(dateTime);
```
should pass-thru and return the DateTime

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+        }
+        if (value is string stringValue && DateTime.TryParse(stringValue, CultureInfo.InvariantCulture, DateTimeStyles.None, out var result))
+        {
+            return DateOnly.FromDateTime(result);
```
should return a DateTime

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+        {
+            return DateOnly.FromDateTime(result);
+        }
+        throw new NotSupportedException($"Cannot convert \"{value}\" into {typeof(DateOnly)}");
```
DateTime

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+
+    public override object ConvertTo(ITypeDescriptorContext? context, CultureInfo? culture, object? value, Type? destinationType)
+    {
+        if (value is DateOnly dateOnly)
```
but... value is a DateTime ???

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+public class DateOnlyToDateTimeConverter : TypeConverter, IExtendedTypeConverter
+{
+    public override bool CanConvertFrom(ITypeDescriptorContext? context, Type? sourceType)
+        => sourceType == typeof(string) || sourceType == typeof(DateTime);
```
should also accept a DateOnly

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+            {
+                return dateOnly.ToDateTime(TimeOnly.MinValue);
+            }
+        }
```
and it should support destinationType being DateOnly

StephaneDelcroix commented on src/Controls/src/Core/DateOnlyToDateTimeConverter.cs:
```diff
@@ -0,0 +1,54 @@
+        throw new NotSupportedException($"Cannot convert \"{value}\" into {typeof(DateOnly)}");
+    }
+
+    public override object ConvertTo(ITypeDescriptorContext? context, CultureInfo? culture, object? value, Type? destinationType)
```
this is the one that matters for the 2-way binding

StephaneDelcroix reviewed (changes requested):
Its still very early in the morning, and I'm not properly caffeinated, but the implementation feels reversed:
When a property, or a type, is attributed with a TypeConverter, we expect ConvertFrom to convert to the actual type, or type of property. This is used e.g. by Xaml inflators to convert FROM a string to a DateTime in this sample
ConvertTo takes the value of the type (known to be of type DateTime in your example) to whatever you want to convert it back, like a string for serialisation, or the type in your ViewModel (DateOnly)

Also, this affects the xaml inflation, and will be executed every time you set a date or time in XAML. and instead of relying on default handling will invoke a converter. For perf reasons, the converter should provide a compiled counterpart for XamlC (but only supports convert from string). This also requires a Xaml Unit test that tests the behaviour of this using all inflators

symbiogenesis committed:
rename converters

symbiogenesis committed:
pass datetime directly

symbiogenesis committed:
fix exception message

symbiogenesis committed:
handle more destination types

symbiogenesis committed:
support from TimeSpan

symbiogenesis commented:
> Its still very early in the morning, and I'm not properly caffeinated, but the implementation feels reversed: When a property, or a type, is attributed with a TypeConverter, we expect ConvertFrom to convert to the actual type, or type of property. This is used e.g. by Xaml inflators to convert FROM a string to a DateTime in this sample ConvertTo takes the value of the type (known to be of type DateTime in your example) to whatever you want to convert it back, like a string for serialisation, or the type in your ViewModel (DateOnly)
> 
> Also, this affects the xaml inflation, and will be executed every time you set a date or time in XAML. and instead of relying on default handling will invoke a converter. For perf reasons, the converter should provide a compiled counterpart for XamlC (but only supports convert from string). This also requires a Xaml Unit test that tests the behaviour of this using all inflators

Addressed all the comments. Should handle more conversions now. Did not add XAML tests.

rmarinho commented:
> > Its still very early in the morning, and I'm not properly caffeinated, but the implementation feels reversed: When a property, or a type, is attributed with a TypeConverter, we expect ConvertFrom to convert to the actual type, or type of property. This is used e.g. by Xaml inflators to convert FROM a string to a DateTime in this sample ConvertTo takes the value of the type (known to be of type DateTime in your example) to whatever you want to convert it back, like a string for serialisation, or the type in your ViewModel (DateOnly)
> > Also, this affects the xaml inflation, and will be executed every time you set a date or time in XAML. and instead of relying on default handling will invoke a converter. For perf reasons, the converter should provide a compiled counterpart for XamlC (but only supports convert from string). This also requires a Xaml Unit test that tests the behaviour of this using all inflators
> 
> Addressed all the comments. Should handle more conversions now. Did not add the XAMLC counterpart or associated tests.
> 
> Is there perhaps a good example XAMLC converter and associated tests that I could use as a reference?

Heres an example of how to add a test that also runs against xamlc. the boolean tests with and without xamlc enables 

https://github.com/dotnet/maui/blob/main/src/Controls/tests/Xaml.UnitTests/ColorConverter.xaml.cs#L24

rmarinho committed:
Fix api

rmarinho committed:
[controls] Fix the converters

rmarinho committed:
Fix tests

rmarinho commented:
/azp run

StephaneDelcroix reviewed (changes requested):
DateTime and TimeSpan conversions are built-in, not sure wee ned to add a converter to support xOnly, but we might need to extend the support of the built-in conversion

could you please ad some xaml unit tests ?

also, this would need a compiled converter, or it'll slow xaml inflation of all files with dates or times in it

PureWeen reviewed (changes requested):
@symbiogenesis can you rebase and address @StephaneDelcroix 's comments? 

If you don't have time currently just let me know and we can take over
